### ISSUE TITLE
Args vs. State

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#15]
Currently we have an args API, that lets us associate each action with a particular entity (like a post or user). The issue I want to explore here is whether args and "stateful args" should be separated.

Stateful args would include, for example, the current user, the current post, and the current site on multisite.

As you can see, each of these corresponds to a type of entity that will already be registered with the args API. We don't want to have to register that type of entity under a second name just to make it stateful, because that would result in all of the children of the stateless entity (atts and relationships) not being automatically registered with the stateful entity.

What we really need is just a way to retrieve the current value for each type of entity (that can be stateful) from the current state.

There are two basic approaches to doing that. One would be to add an interface like this, and let each entity decide whether it wants to implement it:

``` php
interface WordPoints_Hook_Arg_Entity_Stateful {
    public function set_the_value_from_state();
}
```

The other way would be to store the information about whether an entity was stateful elsewhere, instead of in the entity class declaration.

Because the API is composible, both of the options are really quite similar from the perspective of getting the statefulness from the entity object. To make a nonstateful entity stateful, we would just subclass it and register the subclass in its place.

However, the two are significantly different when approached from the other direction, when, for example, we just want a list of stateful entities. So we need to determine whether that will be important before we decide which of these routes to take.

**Comment by JDGrimes** [comment]
I currently see no reason why we would need a list of all of the stateful args. However, lets consider how they'd be used in the rest of the API before we make a final decision.

To map out the rest of the API, we need to think about where we'll need to be retrieving an entity value from the state.

Obviously, this would begin when an action is fired. At that point, we may want to have the main entity be from the current state instead of one of the hook args.

An example of this would be the "user visit" action. We would want to associate this action with the current user. Depending what action we hook into, the current user might not be one of the hook args.

Next, after the action is triggered, all of the associated events will be fired. The events don't really touch the entity passed to them, but they do pass it on to the extensions, and, finally, the reactor.

Lets consider how the Conditions extension might use the stateful args. It is certainly possible that a user might want to set a condition on a stateful arg. In order for the condition object to handle that, it will need to be able to get the value of that arg. Currently, the conditions are all under a hierarchical array with only the main entity at the highest level. Other entities could be added there as well, however, we would run into conflicts between the entity from the state and the entity from the hook. That might not seem like a big problem, because in many cases the two would have the same value anyway. But there are cases where that isn't so. For example, when a user is editing another user.

To avoid this, we need to be able to differentiate between action arg entities and state entities at the top of an hierarchy. This is true everywhere that hierarchies are used, not just for Conditions.

The question, then, is how should we do this?

One option would be to put them in separate hierarchies altogether, however, that may overcomplicate things.

The idea that I have toyed with in the last few days is to register a special entity that represents the state. Then all of the stateful entities would be registered as relationships tying the current state to the regular entity. All of the logic for retrieving the current value from the state is then contained in the relationship.

This solution seemed elegant to me, because it used the existing API, it did not require modifying the class declaration of an entity to make it stateful, all of the stateful entities would be in a list, and the children of the entities would not have to be registered twice.

It works great for pulling stateful entities from the hierarchy, but we run into an issue when using the stateful args in place of action arguments. ~~~The entity associated with an action must be at the top of the hierarchy, so it would now be the entity for the current state. For example, instead of the entity associated with the Visit action being a User, as above, it would be the State entity instead.~~~ This is where I was getting hung up. The truth is that it just needs to be an entity, any entity, because **every entity is at the top of the hierarchy**. So the action can return the stateful entity if it wants to.

The problem then is that we now have the entity in the hierarchy twice. That is, under `state -> entity` and `entity`. So for the Visit action, for example, we would have the User and the Current User both in displayed in the UI. For that reason, we'd need to be able to identify when the event was stateful, so that we could avoid that. For all other cases, the stateful entity and the action entity may be different, so it makes sense to display them both. Even here, it is mostly a UI problem, since the two places in the hierarchy will have the same value in the API when the action fires.

Note that I said "we'd need to be able to identify when the event was stateful" above, though. The action is what would actually be stateful, but it is decoupled from the event. As long as it returns the entity value that the event is expected to be associated with, why should it make any difference how that value is retrieved? The event, and all other pieces of the API, should be able to ignore this. So we wouldn't have "stateful events" at all, just stateful actions.

But the actions aren't currently exposed to the UI in any way, which means that there would be no way to determine whether an event was hooked to a stateful action. However, if the other pieces of the API can function without that info on the back end, shouldn't they be able to in the UI as well?

But that issue of redundancy still bothers me. In the API, possibly we could solve it by creating the hierarchy with the State arg instead of the User arg, while still marking the User arg as the main one. That would solve the problem, however, it wouldn't fix the issue in the UI, unless we expose the action data to it. I guess it might not be possible to solve this and maintain a decoupled API without doing that, though.

That solution wouldn't maintain a decoupled API though, because the hierarchies saved in any reactions would break if the stateful action was swapped for a non-stateful one. That is true whenever an action is swapped for another action with a different type of entity. But in this case we'd be splitting the action entity from the top of the hierarchy. I don't think that is a good "solution".

RE UI: all we need to know is whether the action entity and state entity are the same value. Is that a per-action or per-event thing? Really, there are probably lots of actions that we could hook into for any given event, but that doesn't change whether the action entity and state entity have the same value. So it is really the event that is stateful, whatever the action is. So the visit event would be stateful. In that case, the event knows it will get the same value from action whether it comes from the state or the action args. So we would be justified in marking the event itself as stateful.

As far as the redundancy in the API, we could have the stateful events actually reconstruct the hierarchy to have the State entity at the top. That way there would be consistency in the top of the hierarchy no matter what action was used, while there would still be no redundancy in the API. However, this is a bit of a hack.

Let's ask ourselves why we need the action to be returning the User entity instead of the State entity at all? Can't the event just be associated with the State entity?

The answer is yes, however, this causes issues when we come to the reactor. Because we can't reverse a State event (at least by default). So we need to ask ourselves whether it makes sense to reverse stateful events in the first place.

For the Visit event I was going to have it reverse if the visiting user was deleted. (See https://github.com/WordPoints/hooks-api/issues/2#issuecomment-105310458) However, I don't think that is really necessary. It could be added later if desired.

My Proposal:
- [x] Make reversal actions optional.
- [x] Don't register a reverse action for the Visit event by default.
- [x] Make the entity for the Visit event the State entity.
- [x] Stateful events would be easily identified in the UI by the fact that their entity was the State entity.

**Comment by JDGrimes** [comment]
We could now safely expose the state entities in the UI. The main State entity is the main entity for the Visit hook, and there are no problems there. For events which are associated with other entities, we can expose the stateful entities alongside the other entities. This is useful, for example, when we want to award a moderator for approving a comment; we would award the Current User.

What is lacking here, though, is user education. How will users know when to use the stateful entities, and when not to? Specifically, how do we communicate to the user what role each of the stateful entities plays in a given event? This is important, because there are pitfalls.

One pitfall, for example, is when selecting the Current User to award points to when a post is published. The user might assume that this will always be the same as the post author, but it won't. The author of the post may only be a contributor, and the post won't be published until it has been reviewed by another user with higher capabilities. Again, this is useful, if we want to award the moderator. However, I'm not sure how quickly users will catch on.

And there are other, more tedious pitfalls, especially related to the Current Post.

Will users find learning this through experience too painful? How can we decrease that possibility?

One thought is to disable by default the inclusion of the stateful entities with the main entity for non-stateful events. We could then allow it to be re-enabled, either universally or per-event, via a filter or something. This would let power-users take advantage of this feature, without confusing less-savvy users. And because of that granularity, we could enable certain stateful entities to be made available with certain events by default, even while it was disabled for all of the others.

**Comment by JDGrimes** [comment]
## Summary

Wow, I was really rambling some of the time above. :-)

The conclusion is as follows:
- The stateful entities should be hidden by default.
- They should be added to each event individually.

This basically means that they won't be all registered in one place. However, I think that is OK. The relationship of a given entity to the event will vary from one event to another anyway.

This also means that the Events won't be marked as stateful/non-stateful. They no longer need to be, because that info isn't needed to properly display the entities related to an event in the API.

To fully do this, we need to do a few more things:
- [x] Remove the State entity.
- [x] Provide a different entity for the Visit event.
- [x] Allow new entities to be registered for an event.

**Comment by JDGrimes** [comment]
Just a note that this ties right in with some discussion on https://github.com/WordPoints/hooks-api/issues/13#issuecomment-142439749 (which really should have been put in its own issue).

**Comment by JDGrimes** [comment]
## Arg Aliases

We still have to decide how to handle two issues:
1. Multiple args of the same type for a single event.
2. Child args needing to be registered for the stateful args.

I think the obvious solution to these two dilemmas is to introduce the concept of arg aliases into the API. So then the current user arg would be an alias of the user arg.

**Comment by JDGrimes** [comment]
## Actions or Events?

Before we can create the API to register new entities for an event, we have to settle the question as to whether the entities should be registered for the event, or for the action.

### On the Action

Registering the args on the Action has several benefits.
- The Action is a reusable piece of API so the entities only need to be registered once across several events.
- How to retrieve the arg value is usually tied to the action arguments anyway.

### On the Event

Registering on the Event also has its benefits.
- The entities for an event can be retrieved without having to retrieve its action.
- We don't have to worry about parity of entities across actions (when an event uses several actions, or actions are switched out).

### Discussion

Reusability of the actions is actually seriously in question. For the most part, they will be used once, and only once. Even if they will be reused occasionally, not having to re-register args is nothing more than sugar at best, and could actually be a problem if the context of the args is different across the different events. It isn't an argument in favor of actions.

The fact that the retrieval of entity values often relies on the action arguments is a more persuasive argument. But is it conclusive?

We must weigh the arguments for the other side, which essentially are that the API is less "breakable" when args are not dependent on possibly-ephemeral Action-Event associations. The problem is that the sword we used above cuts both ways. The actions will seldom be switched-out, if ever. So it isn't a strong objection.

The thing that does give some strength to it is the fact that some Events can be hooked to multiple Actions, and parity across them is an absolute necessity for the reactions to work. Of course, that is pretty obvious since all of the actions would have to be doing the same thing to the same type of entity. We just expect that the affected entity would be one of the args of the action. And I don't think that that is a crippling assumption for us to make within our API. The only thing that would make me question it would be non-`fire` actions (such as reverse/spam/etc.). But I think that even there, when used properly, we wouldn't really ever need to hook the event to an action that didn't have the entity as one of its args. It's just common sense.

The strongest argument on either side, then is the fact that the retrieval of an Entity value will be decided by the action arguments. So it makes sense that the Action object should be involved in supplying that value.

### Conclusion

Perhaps there is room for compromise here. The entities could be registered on the Event, but some of the values could be supplied by the Action object, while other values would be pulled from elsewhere. This could give us the best of both worlds, in that we could still retrieve all of the entity types for an Event without having to check its actions.

**Comment by JDGrimes** [comment]
## Alias Syntax

I think we should introduce a specified syntax for alias slugs. This would be beneficial for several reasons:
- Consistency is always a good thing.
- It could help us avoid conflicts between alias names and entity names.
- It could even make it possible to determine the entity type directly from the alias slug.

To accomplish all of these, I've been toying with the idea of alias slugs in this format:

```
slug         = *( ALPHA / DIGIT / "_" )
entity_slug  = slug
entity_array = entity_slug "{}"
entity_alias = slug ":" entity_slug
```

This basically turns aliases into entity modifiers: `current:user`, `previous:password`, etc. I think this works rather well.

The one unsettled question is whether we need to be able to alias entity arrays. We don't right now, because all top-level entities in the hierarchy must be single entities, not entity arrays. But if we somehow add that in the future, our syntax would still work rather well.

**Comment by JDGrimes** [comment]
> Provide a different entity for the Visit event.

In the end, this meant that no entities for the Visit event would come from the action (at least for now, since we're using the `wp` action). But we'll supply the `current:user` entity (and probably others like `current:post` and `site`) from the current state.

**Comment by JDGrimes** [comment]
I think this question has been settled. I'm not sure that the code currently works, but if we run into other minor details that need some work, we can open new issues for them.
